### PR TITLE
Add TCP custom op builder helper

### PR DESCRIPTION
**Pull request by srinathava**
Add a utility to aid in converting torch ops to `tcp.custom_op`

### CODE REVIEW

File: lib/Conversion/TorchToTcp/TcpCustomOpBuilder.cpp
Comment (navahgar) [inline lib/Conversion/TorchToTcp/TcpCustomOpBuilder.cpp:19]: It should conform to Tcp_Tensor which is defined here: https://github.com/cruise-automation/mlir-tcp/blob/main/include/mlir-tcp/Dialect/IR/TcpTypes.td#L54. But that should automatically happen with the verifier. So, there is no need to explicit error checking here.

File: include/mlir-tcp/Conversion/TorchToTcp/TcpCustomOpBuilder.h
Comment (navahgar) [inline include/mlir-tcp/Conversion/TorchToTcp/TcpCustomOpBuilder.h:56]: Do you have any strong reasons to have these as methods on a class as compared to stand-alone functions?

Here are the advantages I see with stand-alone functions:
* These functions can be used through this conversion code (not just the custom op conversions) without instantiating any objects. We don't have to keep track of the object lifetime in that case.
* Builder is already over-loaded in MLIR. Adding one more here seems confusing.
* That is the approach followed in MLIR AFAICT. So, there will be uniformity.

So, I strongly recommend moving to stand-alone functions.

Comment (navahgar) [reply include/mlir-tcp/Conversion/TorchToTcp/TcpCustomOpBuilder.h:56]: @sjain-stanford Can you think of other reasons for going with stand-alone functions?

**Review by navahgar — COMMENTED** [review]
Thanks for taking this up.

File: include/mlir-tcp/Conversion/TorchToTcp/TcpCustomOpBuilder.h
Comment (srinathava) [reply include/mlir-tcp/Conversion/TorchToTcp/TcpCustomOpBuilder.h:56]: The reason I went with methods is that as we call them, we are accumulating the state of the built-up tcp.custom_op's operands etc. into `operandNames`, `operands`, `conversionResult`, `attrs`. Without these things being methods on some class, all of that state will have to be hoisted into the callee function and passed in. At that point, the code basically becomes "classes in C" as opposed to a C++ class. At least I cannot immediately imagine writing the callee-side code as cleanly without this encapsulation. See for instance, this bit:

```cpp
    builder.addOperand("self", adaptor.getSelf());
    builder.addOperand("index", adaptor.getIndex());
    builder.addIntAttr("axis", op.getDim());

    return builder.replace();
```

Without the class, would we have to declare the members of the class in that function (operandNames etc.) and then pass these in by reference to these (now free) functions? How would that be really different?

File: include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.h
Comment (sjain-stanford) [inline include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.h:30]: I believe this should say `non-tensor operands`, since we keep all tensor operands as-is.

File: BUILD
Comment (sjain-stanford) [inline BUILD:198]: Prefer using the existing `lib/Conversion/TorchToTcp/Utils.cpp` instead?

File: lib/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.cpp
Comment (sjain-stanford) [inline lib/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.cpp:57]: remove TODO as we last discussed

File: BUILD
Comment (navahgar) [reply BUILD:198]: One alternative is to create a `TorchToTcpCustomOpUtils.h` and `.cpp`, to hold those that are specific to custom op cases. Wdyt?

**Review by navahgar — COMMENTED** [review]
LGTM. I'll let @sjain-stanford approve once his comment is addressed.

File: BUILD
Comment (sjain-stanford) [reply BUILD:198]: Given that `TorchToTcpCustomOp` lib  is not factored as a separate conversion target (i.e. it reuses `TorchToTcp` [target](https://github.com/cruise-automation/mlir-tcp/blob/main/BUILD#L188)), and we already have some utilities in `Utils.cpp` that are shared by both (e.g. `addPatternIfOpInConvertTorchOpsSet`), it might make sense to keep them colocated. There are organizational benefits in doing so, like making future refactorings for code reuse easier.

File: lib/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.cpp
Comment (sjain-stanford) [inline lib/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.cpp:27]: Do we need the early return here too (like in the other methods)?
```
  if (conversionResult.failed())
    return;
```

File: lib/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.cpp
Comment (sjain-stanford) [inline lib/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.cpp:19]: Do we need the early return here too (like in the other methods)?
```
  if (conversionResult.failed())
    return;
```

File: include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.h
Comment (srinathava) [reply include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.h:30]: I'll change "tensor operands" to just "operands". My understanding is that a torch OP _can_ have a tensor operand which we merely insist is a constant.

File: BUILD
Comment (srinathava) [reply BUILD:198]: Interesting... In a previous life, we preferred NOT having files like "utils" etc. because they would become a general catch-all of eventually unrelated things and preferred more standalone headers (ideally a header per class decl.) to make dependencies in the C++ file much more granular.

I have no strong preferences here. @sjain-stanford , did you mean to just put the implementation (cpp file) into `Utils.cpp` or also the helper class into `Utils.h`?

Comment (sjain-stanford) [reply BUILD:198]: I'm going by patterns from the `llvm-project`/`torch-mlir` codebases (which we try to mimic in TCP). See some examples below:
- https://github.com/llvm/torch-mlir/blob/main/lib/Conversion/Utils/Utils.cpp
- https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SCF/Utils/Utils.cpp
- https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Tosa/Utils/ConversionUtils.cpp

> did you mean to just put the implementation (cpp file) into Utils.cpp or also the helper class into Utils.h?
Let's move the class interface to `Utils.h` as well (this is already included for use by the conversion patterns).

File: include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.h
Comment (sjain-stanford) [reply include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.h]: >  My understanding is that a torch OP can have a tensor operand which we merely insist is a constant.

FWICT we always preserve tensor operands as tensors. Yes, scalar types get converted to attributes, sometimes even if they're not constants. @navahgar Can you think of a case otherwise?

[TorchTypes.td](https://sourcegraph.com/github.com/llvm/torch-mlir@main/-/blob/include/torch-mlir/Dialect/Torch/IR/TorchTypes.td) has a clear separation between tensor and non-tensor (aka scalar) dtypes. The regular types like `Torch_IntType` or `Torch_FloatType` are explicitly modeled after python ints/floats and it notes `This type is not used for modeling tensor dtypes`. 
```
def Torch_IntType : Torch_Type<"Int", "int"> {
  let summary = "Torch IntType";
  let description = [{
    The integer type used to model the Python `int` type in TorchScript.
    TorchScript itself models this type as a 64-bit signed integer.

    Note: This type is not used for modeling tensor dtypes.
  }];
}
```
Please correct me if I'm missing something.

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sjain-stanford) [inline lib/Conversion/TorchToTcp/Utils.cpp:410]: Can we make this conditional on `indexTensors.size() > 1` (to avoid the single element list case having a redundant `_0` suffix)

**Review by sjain-stanford — APPROVED** [review]
LGTM. Thanks for the thoughtful refactoring - this significantly simplifies writing and maintaining TCP custom op conversions.

File: include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.h
Comment (navahgar) [reply include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.h]: > > My understanding is that a torch OP can have a tensor operand which we merely insist is a constant.
> 
> FWICT we always preserve tensor operands as tensors. 

That's right. Tensor operands are always inputs to the ops. At this point during the lowering, I can't think of a case where we insist this to be constants.

But later in the compilation, we have some cases, like in the QuantSim ops, where we insist that they are constants.

> Yes, scalar types get converted to attributes, sometimes even if they're not constants.

One clarification here: they have to be constants to be converted to attributes. We probably error out if they are not constants in such cases.

**Pull request merged**
